publisher/publisher_test.go

```diff
@@ -0,0 +1,413 @@
+	err = testReceipt.CheckSignature()
+	test.AssertError(t, err, "Invalid signature check failed")
+}
+
+func TestSubmitToCT(t *testing.T) {
+	server := logSrv()
+	defer server.Close()
+	port, err := getPort(server)
+	test.AssertNotError(t, err, "Failed to get test server port")
+
+	intermediatePEM, _ := pem.Decode([]byte(testIntermediate))
+
+	pub, err := NewPublisherImpl(CTConfig{Logs: []LogDescription{LogDescription{URI: fmt.Sprintf("http://localhost:%d", port)}}, SubmissionBackoffString: "0s", BundleFilename: issuerPath})
+	pub.issuerBundle = append(pub.issuerBundle, base64.StdEncoding.EncodeToString(intermediatePEM.Bytes))
+	pub.SA = &mocks.MockSA{}
+	test.AssertNotError(t, err, "Couldn't create new Publisher")
+
+	leafPEM, _ := pem.Decode([]byte(testLeaf))
+	leaf, err := x509.ParseCertificate(leafPEM.Bytes)
+	test.AssertNotError(t, err, "Couldn't parse leafPEM.Bytes")
+
+	log.Clear()
+	err = pub.SubmitToCT(leaf.Raw)
+	test.AssertNotError(t, err, "Certificate submission failed")
+
+	// No Intermediate
+	pub.issuerBundle = []string{}
+	log.Clear()
+	err = pub.SubmitToCT(leaf.Raw)
+	test.AssertNotError(t, err, "Certificate submission failed")
+}
+
+func TestGoodRetry(t *testing.T) {
+	server := retryableLogSrv(1, nil)
+	defer server.Close()
+	port, err := getPort(server)
+	test.AssertNotError(t, err, "Failed to get test server port")
+
+	intermediatePEM, _ := pem.Decode([]byte(testIntermediate))
+
+	pub, err := NewPublisherImpl(CTConfig{
+		Logs: []LogDescription{LogDescription{URI: fmt.Sprintf("http://localhost:%d", port)}},
+		SubmissionBackoffString: "0s",
+		BundleFilename:          issuerPath,
+		SubmissionRetries:       1,
+	})
+	pub.issuerBundle = append(pub.issuerBundle, base64.StdEncoding.EncodeToString(intermediatePEM.Bytes))
+	pub.SA = &mocks.MockSA{}
+	test.AssertNotError(t, err, "Couldn't create new Publisher")
+
+	leafPEM, _ := pem.Decode([]byte(testLeaf))
+	leaf, err := x509.ParseCertificate(leafPEM.Bytes)
+	test.AssertNotError(t, err, "Couldn't parse leafPEM.Bytes")
+
+	log.Clear()
+	err = pub.SubmitToCT(leaf.Raw)
+	test.AssertNotError(t, err, "Certificate submission failed")
+}
+
+func TestFatalRetry(t *testing.T) {
+	server := retryableLogSrv(1, nil)
+	defer server.Close()
+	port, err := getPort(server)
+	test.AssertNotError(t, err, "Failed to get test server port")
+
+	intermediatePEM, _ := pem.Decode([]byte(testIntermediate))
+
+	pub, err := NewPublisherImpl(CTConfig{
+		Logs: []LogDescription{LogDescription{URI: fmt.Sprintf("http://localhost:%d", port)}},
+		SubmissionBackoffString: "0s",
+		BundleFilename:          issuerPath,
+	})
+	pub.issuerBundle = append(pub.issuerBundle, base64.StdEncoding.EncodeToString(intermediatePEM.Bytes))
+	pub.SA = &mocks.MockSA{}
+	test.AssertNotError(t, err, "Couldn't create new Publisher")
+
+	leafPEM, _ := pem.Decode([]byte(testLeaf))
+	leaf, err := x509.ParseCertificate(leafPEM.Bytes)
+	test.AssertNotError(t, err, "Couldn't parse leafPEM.Bytes")
```

Comment:
From line 282 to here seems to be repeated amongst several tests. Could you factor it out into a setup function with multiple return values?